Copy+Paste, and other time-saving additions

Kamayana opened the pull request:
### Prerequisites
Mark these off with an X in the square brackets.

* [x] This is merging into the `dev` branch
* [x] I have thoroughly tested this feature
* [x] I have aligned to the code standards
* [x] I agree that permission to use these changes will be irreversibly granted to the owner of this repository

### Description
I added new functions to the editor to make it easier to use, especially for large sweeping adjustments.

1. Copy, cut, and paste
    - CTRL+C: copies the selected entities to the clipboard as a JSON string.
    - CTRL+X: deletes and copies the selected entities to the clipboard as a JSON string.
    - CTRL+V: pastes the entities contained in the clipboard's JSON string, if valid. Pasting will not immediately place the entities, instead they will hover under the cursor, acting as if they were being dragged.
        - While stored patterns could be used as a clipboard of sorts, it nevertheless adds an extra step to what could be a simpler process.

2. Select following/preceding/between
    - CTRL+F: selects all entities following the current selection or, if no selection, the playback marker.
    - CTRL+R: selects all entities preceding the current selection or, if no selection, the playback marker.
    - CTRL+T: selects all entities between the currently selected entities. Does nothing without a selection.
        - Making edits to large sections of a track means a lot of scrolling and box selecting. These functions make it much faster to do this.

3. Quick double & halve tempo
    - Hold SHIFT while adding a tempo change: adds a tempo change at half the speed of the current tempo.
    - Hold ALT while adding a tempo change: adds a tempo change at double the speed of the current tempo.
        - Doubling and halving the tempo is by far the most common tempo change I make; it's worth giving a shortcut.
        - I didn't use CTRL for doubling the tempo since Macs will interpret it as a right-click.

4. Adds the Mac's command key as an alternative to CTRL.

5. Adds 'P' as an alternate hotkey to Store Pattern, as most Mac keyboards don't have an INSERT key.

6. Uses raw strings for the pickaxe tool sound code, as the nested templates were throwing off my IDE's code highlighting. This is obviously a minor change, but I tested the pickaxe tool and it found the sounds just fine.

### Potential Issues
- The new description for the tempo change tool is perhaps too long, and there aren't any new translations for it.
- Pasting entities puts them into a "dragging" state despite the mouse button not being held down. This actually doesn't seem to cause any big issues, all I found is that clicking buttons while "dragging" entities feels weird, and that entities can be pasted during presentation mode.
- Select all following/preceding/between does nothing with tracker entities, since they can't be selected or moved in the first place. I wasn't sure about adding that functionality so I left that alone.

chrislo27 committed:
Fix RHRE3.jar references

Kamayana committed:
Add quick double/halve tempo function to tempo change tool

Kamayana committed:
Add Select All command

Kamayana committed:
Add an alternate hotkey for storing patterns

Most Mac keyboards replace the insert key with fn, so an alternate is needed

Kamayana committed:
Add commands for selecting all following/preceding/between selected

Kamayana committed:
Add the command key as a Mac-specific alternative to Control

Kamayana committed:
Remove unnecessary control+shift check

Kamayana committed:
Use playback start as the default for following/preceding selection

Kamayana committed:
Add copy, cut, and paste functionality and key commands

chrislo27 committed:
Merge branch 'dev'

Kamayana committed:
Merge remote-tracking branch 'upstream/dev' into dev

Kamayana committed:
Rebind following/preceding/between commands to use CTRL instead of shift

Kamayana committed:
Add new bindings to README document

Kamayana committed:
Simplify copy/cut/paste functions

Kamayana committed:
Use raw string for the pickaxe tool sounds

To prevent the nested templates from screwing up syntax highlighting

Kamayana committed:
Clean up whitespace

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/toolboks/util/gdxutils/InputUtils.kt:
```diff
@@ -6,6 +6,7 @@ import com.badlogic.gdx.Input
 
 private val inputMap: MutableMap<Int, Boolean> = mutableMapOf()
 private val buttonMap: MutableMap<Int, Boolean> = mutableMapOf()
+private val isMac: Boolean = System.getProperty("os.name").contains("Mac")
```
Redundant. Use `com.badlogic.gdx.utils.SharedLibraryLoader.isMac` instead. See https://github.com/chrislo27/RhythmHeavenRemixEditor/blob/6f2d07df0ff1a3f9643f431923a0bad4982d9346/core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt#L1192 as an example.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/toolboks/util/gdxutils/InputUtils.kt:
```diff
@@ -44,7 +45,7 @@ fun Input.isButtonJustPressed(button: Int): Boolean {
 
 fun Input.isControlDown(): Boolean {
-    return isKeyPressed(Input.Keys.CONTROL_LEFT) || isKeyPressed(Input.Keys.CONTROL_RIGHT)
+    return isKeyPressed(Input.Keys.CONTROL_LEFT) || isKeyPressed(Input.Keys.CONTROL_RIGHT) || isMac && isKeyPressed(Input.Keys.SYM)
```
I disagree with the idea of adding the ⌘ key as a replacement for CTRL. All control-scheme text is already shown as CTRL.

chrislo27 replied:
Please add explicit parentheses around `isMac && isKeyPressed(Input.Keys.SYM)`.

chrislo27 commented on core/assets/localization/default.properties:
```diff
@@ -98,7 +98,7 @@ editor.msg.selectionHint.count.add={0} will be added to the selection
 editor.msg.cannotSplit=[RED]Cannot split entity[] because the split cues would intersect with other cues
 editor.msg.timeSignature=Left/right click to place/delete time signature changes - Scroll to change beats per measure (Hold [CYAN]CTRL[] for ±5) - Hold [CYAN]SHIFT[] and scroll to change beat divisions
-editor.msg.tempoChange=Left/right click to place/delete tempo changes - Drag the ends for a gradual transition  - Scroll to change - Hold [CYAN]CTRL[] for ±5, [CYAN]SHIFT[] for ±0.1
+editor.msg.tempoChange=Left/right click to place/delete tempo changes - Hold [CYAN]ALT[] while clicking to double tempo, hold [CYAN]SHIFT[] to halve it - Drag the ends for a gradual transition - Scroll to change - Hold [CYAN]CTRL[] for ±5, [CYAN]SHIFT[] for ±0.1
```
"Clicking" is ambiguous. My recommended wording is "Hold [CYAN]ALT[] while placing to double the tempo, [CYAN]SHIFT[] to halve it".

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                        // Select following
+                        val selectionMinX: Float = this.selection.minBy { it.bounds.x }?.bounds?.x ?: remix.playbackStart
+                        val newSelection = remix.entities.toList().filter { it.bounds.x >= selectionMinX }
+                        remix.mutate(EntitySelectionAction(this, this.selection, newSelection))
```
Check that this new selection isn't equal (through size checks and `containsAll`) to the existing selection.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                        // Select preceding
+                        val selectionMaxX: Float = this.selection.maxBy { it.bounds.maxX }?.bounds?.maxX ?: remix.playbackStart
+                        val newSelection = remix.entities.toList().filter { it.bounds.maxX <= selectionMaxX }
+                        remix.mutate(EntitySelectionAction(this, this.selection, newSelection))
```
Check that this new selection isn't equal (through size checks and `containsAll`) to the existing selection.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                        val selectionMinX: Float = this.selection.minBy { it.bounds.x }?.bounds?.x ?: 0f
+                        val selectionMaxX: Float = this.selection.maxBy { it.bounds.maxX }?.bounds?.maxX ?: 0f
+                        val newSelection = remix.entities.toList().filter { it.bounds.x >= selectionMinX && it.bounds.maxX <= selectionMaxX }
+                        remix.mutate(EntitySelectionAction(this, this.selection, newSelection))
```
Check that this new selection isn't equal (through size checks and `containsAll`) to the existing selection.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                        remix.mutate(EntitySelectionAction(this, this.selection, newSelection))
+                    } else if (Gdx.input.isKeyJustPressed(Input.Keys.C)) {
+                        // Copy
+                        Gdx.app.clipboard.contents = PatternStoreScreen.entitiesToJson(remix, selection.toList())
```
Check that the selection is not empty.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                        Gdx.app.clipboard.contents = PatternStoreScreen.entitiesToJson(remix, selection.toList())
+                    } else if (Gdx.input.isKeyJustPressed(Input.Keys.X)) {
+                        // Cut
+                        Gdx.app.clipboard.contents = PatternStoreScreen.entitiesToJson(remix, selection.toList())
```
Check that the selection is not empty.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                        this.selection = listOf()
+
+                        updateMessageLabel()
+                    } else if (Gdx.input.isKeyJustPressed(Input.Keys.V)) {
```
Pasting without explicitly dragging is something I have avoided. Note that so far, copying is holding ALT and clicking and dragging on a selection. I will need to review and consider this further.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                        try {
+                            val result = PatternStoreScreen.jsonToEntities(remix, Gdx.app.clipboard.contents)
+
+                            if (result.isEmpty())
```
Return not recommended. Check for an empty result instead.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                            entities = result
+                        } catch (e: Exception) {
+                            e.printStackTrace()
+                            return
```
Return not recommended.

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -913,6 +913,71 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
+                            it.updateInterpolation(true)
+                        }
+
+                        remix.entities.addAll(entities)
```
This action is not reversible (undo/redo).

chrislo27 commented on core/src/main/kotlin/io/github/chrislo27/rhre3/editor/Editor.kt:
```diff
@@ -1550,7 +1615,13 @@ class Editor(val main: RHRE3Application, stageCamera: OrthographicCamera, attach
                         val tr = when (tool) {
                             Tool.TEMPO_CHANGE -> {
-                                TempoChange(remix.tempos, beat, remix.tempos.tempoAt(beat), remix.tempos.swingAt(beat), 0f)
+                                var tempoScale = 1f
```
Simplify, avoid `var`. 
```kotlin
val tempoScale = if (shift && !alt) 0.5f else if (!shift && alt) 2f else 1f
```

chrislo27 commented on docs/README.md:
```diff
@@ -77,6 +77,12 @@ however, read up on the controls first below!
+| **`CTRL+T`** | Select entities between selections |
+| **`CTRL+C`** | Copy selected entities to clipboard |
+| **`CTRL+X`** | Cut selected entities to clipboard |
+| **`CTRL+V`** | Paste entities from clipboard |
```
CTRL+V is under more review, I'll get back to this.

chrislo27 reviewed (changes requested):
Good work on your changes. I have a few requested changes that I've commented on on their respective lines. ~~The only change I will not pass is the Command key as CTRL.~~ (I added a separate comment regarding the Cmd key, but I will approve it now.)

I'm still uncertain about CTRL+V. I know that CTRL+C/X is sort of useless without CTRL+V but it seems odd that you can place a selection without holding the mouse button. I'll think about this more.

Kamayana committed:
Adjustments based on feedback

-Adjust description of tempo doubling/halving
-Add equal/empty selection checks
-Remove unneeded returns
-Simplified tempoScale calculation
-Standardized command key Mac check

Kamayana commented:
Applied most of the requested changes, but there's still the question of how to make the paste function work. I thought of a few alternate methods:

1. Holding V (+CTRL optional) will create a ghost of the clipboard pattern that moves with the cursor. Releasing V will delete the ghost. Clicking while holding V will turn the ghost pattern into a real pattern that's now being dragged. The ghost pattern should have a text overlay that says what's going on.
2. Similar to the previous, CTRL+V will create an _immobile_ ghost pattern under the cursor. If the user clicks, the ghost is made real and dragged. If the user moves the cursor too far away from the origin point, the ghost is deleted automatically. Feels like it'd be more confusing than method number 1.
3. Add a clipboard icon to the toolbar that can be clicked on to create and drag the clipboard pattern, just like clicking on the pattern browser. Could update the icon to show if the clipboard contains a valid pattern or not. Unsure how to make CTRL+V do anything with this method, but this method could also be combined with either of the other two. It might be a little incongruous since none of the other icon buttons create patterns like this.

chrislo27 commented:
Thank you for addressing my concerns.

I would like to prioritize consistency -- so suggestion 3 appeals to me the most.  
Here is my suggestion: a `(Clipboard)` pattern can be displayed in the Stored Patterns section as the first item when a pasteable selection is in the clipboard. `CTRL+V` would simply jump to that item (like how pressing `MOUSE3` on an entity jumps to that one in the pickers). 

Please let me know what you think.

Kamayana commented:
> Thank you for addressing my concerns.
> 
> I would like to prioritize consistency -- so suggestion 3 appeals to me the most.
> Here is my suggestion: a `(Clipboard)` pattern can be displayed in the Stored Patterns section as the first item when a pasteable selection is in the clipboard. `CTRL+V` would simply jump to that item (like how pressing `MOUSE3` on an entity jumps to that one in the pickers).
> 
> Please let me know what you think. I could do the implementation of CTRL+V and merge the rest of your changes too, if you'd like.

I like that idea. If you want to implement it go ahead, I won't be able to put in time for a few days.

chrislo27 commented:
Switching back to the previous location should be trivial with the quick switch (`F`) keybind. I don't think switching back to the previous location automatically is a good idea, though, in case the user wants to paste the clipboard repeatedly.

chrislo27 commented:
As of commit https://github.com/chrislo27/RhythmHeavenRemixEditor/commit/c27bd96b5524d27b8124dbccbb8d449d7f1f4382 this has been implemented. Thanks again for the other feature implementations!

![screenshot](https://user-images.githubusercontent.com/6299069/56686013-0a8b5680-6688-11e9-89d2-a8136eb0f206.png)

Kamayana commented:
Glad to help. If I did anything in the future, would you prefer to have separate pull requests for each feature rather than the singular hodgepodge this one was? Even when it's just a 2-line change like the CMD-key addition? This is my first time actually contributing to a repo so I'm unsure what the etiquette is.

chrislo27 commented:
So long if you post a changelog (like you did) then I'm okay with one larger pull request that encompasses multiple changes.
